packages/application-generic/src/services/in-memory-provider/cache-in-memory-provider.service.ts

```diff
@@ -0,0 +1,105 @@
+import { Logger } from '@nestjs/common';
+
+import { InMemoryProviderService } from './in-memory-provider.service';
+import {
+  InMemoryProviderEnum,
+  InMemoryProviderClient,
+  ScanStream,
+} from './types';
+
+import { GetIsInMemoryClusterModeEnabled } from '../../usecases';
+
+const LOG_CONTEXT = 'CacheInMemoryProviderService';
+
+export class CacheInMemoryProviderService {
+  public inMemoryProviderService: InMemoryProviderService;
+  public isCluster: boolean;
+  private getIsInMemoryClusterModeEnabled: GetIsInMemoryClusterModeEnabled;
+
+  constructor() {
+    this.getIsInMemoryClusterModeEnabled =
+      new GetIsInMemoryClusterModeEnabled();
+
+    const provider = this.selectProvider();
+    this.isCluster = this.isClusterMode();
+
+    const enableAutoPipelining =
+      process.env.REDIS_CACHE_ENABLE_AUTOPIPELINING === 'true';
+
+    this.inMemoryProviderService = new InMemoryProviderService(
+      provider,
+      this.isCluster,
+      enableAutoPipelining
+    );
```

Comment:
With these changes this parent service and WorkflowInMemoryProviderService, will drive down which provider will be set, instead of delegating that responsibility to the InMemoryProviderService.